src/theory/strings/infer_proof_cons.cpp

```diff
@@ -0,0 +1,1043 @@
+/*********************                                                        */
+/*! \file infer_proof_cons.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2019 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Implementation of inference to proof conversion
+ **/
+
+#include "theory/strings/infer_proof_cons.h"
+
+#include "expr/skolem_manager.h"
+#include "options/smt_options.h"
+#include "options/strings_options.h"
+#include "theory/builtin/proof_checker.h"
+#include "theory/rewriter.h"
+#include "theory/strings/regexp_operation.h"
+#include "theory/strings/theory_strings_utils.h"
+
+using namespace CVC4::kind;
+
+namespace CVC4 {
+namespace theory {
+namespace strings {
+
+InferProofCons::InferProofCons(context::Context* c,
+                               ProofNodeManager* pnm,
+                               SequencesStatistics& statistics)
+    : d_pnm(pnm), d_lazyFactMap(c), d_statistics(statistics)
+{
+  Assert(d_pnm != nullptr);
+}
+
+void InferProofCons::notifyFact(const InferInfo& ii)
+{
+  Node fact = ii.d_conc;
+  Trace("strings-ipc-debug")
+      << "InferProofCons::notifyFact: " << ii << std::endl;
+  if (d_lazyFactMap.find(fact) != d_lazyFactMap.end())
+  {
+    Trace("strings-ipc-debug") << "...duplicate!" << std::endl;
+    return;
+  }
+  Node symFact = CDProof::getSymmFact(fact);
+  if (!symFact.isNull())
+  {
+    if (d_lazyFactMap.find(symFact) != d_lazyFactMap.end())
+    {
+      Trace("strings-ipc-debug") << "...duplicate (sym)!" << std::endl;
+      return;
+    }
+  }
+  std::shared_ptr<InferInfo> iic = std::make_shared<InferInfo>(ii);
+  d_lazyFactMap.insert(ii.d_conc, iic);
+}
+
+void InferProofCons::convert(Inference infer,
+                             bool isRev,
+                             Node conc,
+                             const std::vector<Node>& exp,
+                             ProofStep& ps,
+                             TheoryProofStepBuffer& psb,
+                             bool& useBuffer)
+{
+  // by default, don't use the buffer
+  useBuffer = false;
+  // Must flatten children with respect to AND to be ready to explain.
+  // We store the index where each flattened vector begins, since some
+  // explanations are grouped together using AND.
+  size_t expIndex = 0;
+  std::map<size_t, size_t> startExpIndex;
```

Comment:
Why isn't this a `vector<size_t>`?

Reply:
It can be, this must have been leftover from a previous design.